lua/hbac/setup.lua

```diff
@@ -1,18 +1,75 @@
 local command = require("hbac.command")
-local state   = require("hbac.state")
+local state = require("hbac.state")
 
-local M = {}
+local M = {
+	CONSTANTS = {
+		AUGROUP_AUTO_CLOSE = "hbac_autoclose",
+	},
+	opts = {
+		threshold = 10,
+		autoclose = true,
+	},
+}
 
 local id = vim.api.nvim_create_augroup("hbac", {
-	clear = false
+	clear = false,
 })
 
-M.setup = function(opts)
-	opts = opts or {
-		autoclose = true,
-		threshold = 10
-	}
+M.setup_autoclose = function()
+	state.autoclose_enabled = true
+	vim.api.nvim_create_autocmd({ "BufEnter" }, {
+		group = vim.api.nvim_create_augroup(M.CONSTANTS.AUGROUP_AUTO_CLOSE, { clear = true }),
+		pattern = { "*" },
+		callback = function()
+			local current_buf = vim.api.nvim_get_current_buf()
+			local buftype = vim.api.nvim_buf_get_option(current_buf, "buftype")
+			-- if the buffer is not a file - do nothing
+			if buftype ~= "" then
+				return
+			end
+
+			local buffers = vim.tbl_filter(function(buf)
+				local name = vim.api.nvim_buf_get_name(buf)
+				local listed = vim.api.nvim_buf_get_option(buf, "buflisted")
+				local windows_with_buf = vim.fn.win_findbuf(buf)
+				if name ~= "" and listed and #windows_with_buf == 0 then
+					return true
+				end
+				return false
+			end, vim.api.nvim_list_bufs())
+			local num_buffers = #buffers
+			if num_buffers <= M.opts.threshold then
+				return
+			end
+
+			local buffers_to_close = num_buffers - M.opts.threshold
+
+			-- Buffer sorted by current > pinned > others (by bufnr)
```

Comment:
Cool idea, would it be more performant if we sorted the list on every pin? 🤔 Perhaps nobody has that many open buffers for this to become a problem xD

Reply:
I tested it with ~100 open buffers to no noticeable impact. I believe it shouldn't be a problem unless someone: disables autoclose, opens thousands of buffers and then enables autoclose or sets the buffer threshold absurdly high.